### DISCUSSION TITLE
Why I can't declare a enum/class/struct in a covariant/contravariant interface?

### DISCUSSION THREAD

**Discussion by ghost** [dotnet/csharplang#8699] [Q&A]
Just questions, why I can't declare a enum/class/struct in a covariant/contravariant interface?
And why I can declare an interface that not covariant/contravariant interface in a covariant/contravariant interface?
What is the difference between these two?

```
public interface I1<out T1>
{
    public enum MyEnum { } // CS8427: Enums, classes, and structures cannot be declared in an interface that has an 'in' or 'out' type parameter.

    public class MyClass { } // CS8427: Enums, classes, and structures cannot be declared in an interface that has an 'in' or 'out' type parameter.

    public struct MyStruct { } // CS8427: Enums, classes, and structures cannot be declared in an interface that has an 'in' or 'out' type parameter.

    public interface I2<T2> { } // ok

}
```

**Comment by jilles-sg** [comment]
Internally, types nested in a generic type have the outer type's generic parameters in addition to their own. Therefore, nesting an enum/class/struct in a covariant/contravariant interface would result in a covariant/contravariant enum/class/struct, which is not allowed.

A possible workaround is to move the nested types to a superinterface:
```csharp
public interface I1Base
{
    public enum MyEnum { }

    public class MyClass { }

    public struct MyStruct { }
}

public interface I1<out T1> : I1Base
{
    public interface I2<T2> { }
}
```
It will not be possible to use the covariant/contravariant type parameters in the nested enum/class/struct.

Said another way, `T1` is in scope to be used in `MyClass` and `MyStruct`. In which contexts is this valid and what is the relationship between, for example, `I1<string>.MyClass` and `I1<object>.MyClass`? If the answer is "it is never valid" then the workaround can be used.
